flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java

```diff
@@ -156,16 +165,40 @@ public void unregisterUnusedState(long lowestCheckpointID) {
                         subsumed.add(entry.stateHandle);
                     }
                     it.remove();
+                } else {
+                    markCheckpointInUseAction.accept(entry.createdByCheckpointID);
                 }
             }
         }
-
         LOG.trace("Discard {} state asynchronously", subsumed.size());
         for (StreamStateHandle handle : subsumed) {
             scheduleAsyncDelete(handle);
         }
     }
 
+    @Override
+    public void unregisterUnusedState(long lowestCheckpointID) {
+        // only used for some tests
+        unregisterState(lowestCheckpointID, (x) -> {});
+    }
+
+    @Override
+    public void unregisterUnusedStateAndCheckpoint(
+            long lowestCheckpointID,
+            CheckpointsCleaner checkpointsCleaner,
+            Runnable postCleanUp,
+            @Nullable Executor executor) {
+        Set<Long> checkpointInUse = new HashSet<>();
+        unregisterState(lowestCheckpointID, id -> checkpointInUse.add(id));
+        if (executor != null) {
+            checkpointsCleaner.cleanSubsumedCheckpoints(
+                    lowestCheckpointID, checkpointInUse, postCleanUp, executor);
+        } else {
+            checkpointsCleaner.cleanSubsumedCheckpoints(
+                    lowestCheckpointID, checkpointInUse, postCleanUp, asyncDisposalExecutor);
```

Comment:
WDYT about returning checkpointsInUse from this function (`unregisterUnusedState`) and calling `checkpointsCleaner.cleanSubsumedCheckpoints` outside (i.e. in `CompletedCheckpointStore`)?

That way, `SharedStateRegistry` is kept  unaware of checkpoints cleanup.

Reply:
That makes sense to me, fixed